include/seqan3/io/sequence/sequence_file_in.hpp

```diff
@@ -1,175 +1,911 @@
+    }
 
-    template <typename seqs_type, typename ids_type, typename quals_type, size_t max_records = 0>
-        requires container_concept<typename seqs_type::value> &&
-                 container_concept<typename ids_type::value> &&
-                 container_concept<typename quals_type::value>
-    void read(seqs_type && seqs, ids_type && ids, quals_type && quals);
+    //!\copydoc get
+    template <typename t>
+    friend auto & get(sequence_file_in & file)
+    {
+        file.read_columns();
+
+        return std::get<t>(file.columns_buffer);
+    }
+    //!\}
 
     /* options */
+    //!\brief The options type defines various option members that influence the behaviour of all or some formats.
     struct options_type
     {
-        // post-processing filters that operate on buffer before assignment to out-value
-        std::function<void(std::string &)>  sequence_filter = [] (std::string & seq)  {};
-        std::function<void(std::string &)>   id_filter = [] (std::string & id)   {};
-        std::function<void(std::string &)> qual_filter = [] (std::string & qual) {};
+        //!\brief Read the ID string only up until the first whitespace character.
+        bool truncateID;
     };
+    //!\brief The options are public and its members can be set directly.
     options_type options;
 
 protected:
-    ~sequence_file_in() = default;
+    //!\privatesection
+    /*!\name Data buffers
+     * \{
+     */
+    //!\brief Buffer for a single record.
+    record_type                     record_buffer;
+    //!\brief Buffer of the entire file in columns.
+    columns_type                    columns_buffer;
+    //!\}
 
-private:
-    /* file format */
+    //!\brief Path of the file that the stream operates on.
     std::string file_name;
+
+    //!\brief The stream we are reading from.
     stream_type stream;
-    valid_formats format;
 
-    /* private functions */
-    void select_decompression(std::string const & compress_ext);
-    template <size_t index>
-    void assign_format(std::string const & ext);
+    //!\brief Type of the format, an std::variant over the `valid_formats`.
+    using format_type = detail::transfer_template_args_onto_t<valid_formats, std::variant>;
+    //!\brief The actual std::variant holding a pointer to the detected/selected format.
+    format_type format;
 
-    /* buffers */
+    //!\brief Tell the format to move to the next record and update the buffer.
+    void read_next_record()
+    {
+        assert(!format.valueless_by_exception());
+        std::visit([&] (sequence_file_format_concept & f)
+        {
+            f.read(stream,
+                   options,
+                   detail::get_or_ignore<field::SEQ>(record_buffer),
+                   detail::get_or_ignore<field::ID>(record_buffer),
+                   detail::get_or_ignore<field::QUAL>(record_buffer),
+                   detail::get_or_ignore<field::SEQ_QUAL>(record_buffer));
+        }, format);
```

Comment:
TODO for later: once single_pass_output iterators are merged, pass output iterators directly; decide between "seq,qual" and "seq_qual" in file, not in format